mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+func NewPool(blockchain BlockChain) *Pool {
+	pool := &Pool{
+		prs:                make(map[chainhash.Hash]*wire.MsgMixPairReq),
+		outPoints:          make(map[wire.OutPoint]chainhash.Hash),
+		pool:               make(map[chainhash.Hash]entry),
+		orphans:            make(map[chainhash.Hash]*orphan),
+		orphansByID:        make(map[idPubKey]map[chainhash.Hash]mixing.Message),
+		messagesByIdentity: make(map[idPubKey][]chainhash.Hash),
+		latestKE:           make(map[idPubKey]*wire.MsgMixKeyExchange),
+		sessions:           make(map[[32]byte]*session),
+		sessionsByTxHash:   make(map[chainhash.Hash]*session),
+		epoch:              10 * time.Minute, // XXX: mainnet epoch: add to chainparams
+		expireHeight:       0,
+		expireSem:          make(chan struct{}, 1),
+		blockchain:         blockchain,
+		feeRate:            feeRate,
+		params:             blockchain.ChainParams(),
+	}
+	// XXX: add epoch to chainparams
+	if blockchain.ChainParams().Net == wire.TestNet3 {
+		pool.epoch = 3 * time.Minute
+	}
+
+	if u, ok := blockchain.(UtxoFetcher); ok {
+		pool.utxoFetcher = u
+	}
+	return pool
+}
+
+// Epoch returns the duration between mix epochs.
+func (p *Pool) Epoch() time.Duration {
+	return p.epoch
+}
+
+// MixPRHashes returns the hashes of all MixPR messages recorded by the pool.
+// This data is provided to peers requesting initial state of the mixpool.
+func (p *Pool) MixPRHashes() []chainhash.Hash {
+	p.mtx.RLock()
+	hashes := make([]chainhash.Hash, 0, len(p.prs))
+	for hash := range p.prs {
+		hashes = append(hashes, hash)
+	}
+	p.mtx.RUnlock()
+
+	return hashes
+}
+
+// Message searches the mixing pool for a message by its hash.
+func (p *Pool) Message(query *chainhash.Hash) (mixing.Message, error) {
+	p.mtx.RLock()
+	pr := p.prs[*query]
+	e, ok := p.pool[*query]
+	p.mtx.RUnlock()
+	if pr != nil {
+		return pr, nil
+	}
+	if !ok || e.msg == nil {
+		return nil, fmt.Errorf("message not found")
+	}
+	return e.msg, nil
+}
+
+// HaveMessage checks whether the mixing pool contains a message by its hash.
+func (p *Pool) HaveMessage(query *chainhash.Hash) bool {
+	p.mtx.RLock()
+	_, ok := p.pool[*query]
+	if !ok {
+		_, ok = p.prs[*query]
+	}
+	p.mtx.RUnlock()
+	return ok
+}
+
+// MixPR searches the mixing pool for a PR message by its hash.
+func (p *Pool) MixPR(query *chainhash.Hash) (*wire.MsgMixPairReq, error) {
+	var pr *wire.MsgMixPairReq
+
+	p.mtx.RLock()
+	pr = p.prs[*query]
+	p.mtx.RUnlock()
+
+	if pr == nil {
+		return nil, fmt.Errorf("PR message not found")
+	}
+
+	return pr, nil
+}
+
+// MixPRs returns all MixPR messages with hashes matching the query.  Unknown
+// messages are ignored.
+//
+// If query is nil, all PRs are returned.
+//
+// In both cases, any expired PRs that are still internally tracked by the
+// mixpool for ongoing sessions are excluded from the result set.
+func (p *Pool) MixPRs(query []chainhash.Hash) []*wire.MsgMixPairReq {
+	res := make([]*wire.MsgMixPairReq, 0, len(query))
```

Comment:
Not a big deal, but this seems wasteful to do unconditionally here give a `nil` query will just replace it right away.  I imagine the thought was to do it outside of the critical section, but it's probably be cleaner to just do it right about the for loop over the query.

Reply:
removed the preoptimizing prealloc.